go/database/vt/memory/trie/nodes.go

```diff
@@ -0,0 +1,462 @@
+
+package trie
+
+import (
+	"bytes"
+
+	"github.com/0xsoniclabs/carmen/go/database/vt/commit"
+	"github.com/0xsoniclabs/tracy"
+)
+
+// ---- Nodes ----
+
+// node is an interface for trie nodes, which can be either inner or leaf nodes.
+type node interface {
+	get(key Key, depth byte) Value
+	set(Key Key, depth byte, value Value) node
+	commit() commit.Commitment
+
+	// -- parallel commit support --
+
+	// collectCommitTasks requests the node to append tasks required to be
+	// performed to update its commitment. The resulting list should list the
+	// tasks in order of dependencies, i.e., when processing the list from start
+	// to end, tasks producing inputs for other tasks should appear before
+	// tasks consuming those inputs.
+	collectCommitTasks(tasks *[]*task)
+}
+
+// ---- Inner nodes ----
+
+// inner is the type of an inner node in the Verkle trie. It contains an array
+// of 256 child nodes, indexed by one byte of the key.
+type inner struct {
+	children [256]node
+
+	// The cached commitment of this inner node. It is only valid if the
+	// no dirtyChildCommitments bit is set.
+	commitment commit.Commitment
+
+	// --- Commitment caching ---
+	dirtyChildCommitments bitMap                 // < which children have been modified
+	oldChildCommitments   [256]commit.Commitment // < previous commitments of children
+}
+
+func (i *inner) get(key Key, depth byte) Value {
+	next := i.children[key[depth]]
+	if next == nil {
+		return Value{}
+	}
+	return next.get(key, depth+1)
+}
+
+func (i *inner) set(key Key, depth byte, value Value) node {
+	pos := key[depth]
+	next := i.children[pos]
+	if !i.dirtyChildCommitments.get(pos) {
+		i.dirtyChildCommitments.set(pos)
+		if next != nil {
+			i.oldChildCommitments[pos] = next.commit()
+		}
+	}
+	if next == nil {
+		next = newLeaf(key)
+	}
+	i.children[pos] = next.set(key, depth+1, value)
+	return i
+}
+
+func (i *inner) commit() commit.Commitment {
+	if !i.dirtyChildCommitments.any() {
+		return i.commitment
+	}
+	zone := tracy.ZoneBegin("inner::commit")
+	defer zone.End()
+
+	delta := [commit.VectorSize]commit.Value{}
+	for j := range i.children {
+		if i.dirtyChildCommitments.get(byte(j)) {
+			new := i.children[j].commit().ToValue()
+			old := i.oldChildCommitments[j].ToValue()
+			delta[j] = *new.Sub(old)
+		}
+	}
+
+	// Update the commitment of this inner node.
+	i.commitment.Add(commit.Commit(delta))
+	i.dirtyChildCommitments.clear()
+	return i.commitment
+}
+
+func (i *inner) collectCommitTasks(tasks *[]*task) {
+	if !i.dirtyChildCommitments.any() {
+		return
+	}
+
+	// Produce one task for every dirty child.
+	directChildTasks := make([]*task, 0, 256)
+	delta := [commit.VectorSize]commit.Commitment{}
+	for j := range i.children {
+		if i.dirtyChildCommitments.get(byte(j)) {
+			lengthBefore := len(*tasks)
+			i.children[j].collectCommitTasks(tasks)
+			childTasks := (*tasks)[lengthBefore:]
+
+			var child *task
+			numDependencies := 0
+			if len(childTasks) > 0 {
```

Comment:
Isn't this guaranteed to always be true? If child `j` has a dirty commitment, it also has to spawn a task to update the commitment, no?

Reply:
The inner node assumes that every set sent to the child to result in a dirty state. However, leaf-nodes have the optimization that if the value your are setting is the same that there already was, they do not mark themselves as dirty.  In this case, they would not produce a task.

Reply:
Ah yes, makes sense!